### PR TITLE
Allow examples to be library

### LINKED ISSUE #2358: Allow specifying crate-type on examples
I'm working on a crate that is specifically designed to be used in libraries. As such, I want to include some example libraries. It is not currently possible to have cargo build these, as the `crate-type` is ignored in the `[[examples]]` section.

### PR DESCRIPTION
**Pull request by KalitaAlexey**
This PR allows to specify **crate-type** to an example in **Cargo.toml**.
After this PR an example's **crate-type** can be:

* lib
* rlib
* dylib
* proc-macro

Please look at src/cargo/core/manifest.rs:116 because I am not sure whether I done it right.

I haven't added any tests.
I'd like to add them if someone says me how to do that.

Fixes #2358.

### CODE REVIEW

**Comment by alexcrichton** [comment]
Looks good to me, thanks! The line with TODO looks ok to me and the failure on Travis seems relevant (just a style issue).

If you take a look at `tests/*.rs` you can probably just add a new tests to `tests/test.rs`

**Comment by KalitaAlexey** [comment]
@alexcrichton,
I compiled cargo using `cargo build`.
Which command should I run to check styles?

I will add some tests.

**Comment by KalitaAlexey** [comment]
I allowed an example to be built as a library.
Should `cargo test --example` work?

**Comment by KalitaAlexey** [comment]
@alexcrichton,
What should **cargo** do if I run `test --example <some_example_name>`?
I checked it with file:
```rust
fn main() {
    panic!("");
}

#[test]
fn f() {
    panic!("");
}
```
It does nothing.
It writes that it compiles and then nothing.

**Comment by matklad** [comment]
>Which command should I run to check styles?

`./tests/check-style.sh`

This just checks for lines longer than 100 characters, nothing fancy.

**Comment by KalitaAlexey** [comment]
@alexcrichton,
Bot failed because of timeout.

**Comment by alexcrichton** [comment]
Thanks @KalitaAlexey! Could you also add assertions that the relevant files exist (e.g. dylibs and such) and also remove the TODO?

**Comment by KalitaAlexey** [comment]
@alexcrichton,
I will try.

**Comment by KalitaAlexey** [comment]
@alexcrichton,
Should `cargo test` also work?

**Comment by KalitaAlexey** [comment]
I don't know how to check whether a library exists.

**Comment by KalitaAlexey** [comment]
I have found out that cargo uses rustc to determine the prefix and the suffix of a library.
I have no idea how should I expose that.

**Comment by alexcrichton** [comment]
@KalitaAlexey you can use the `existing_file` function to test that a file exists or you could just read the directory and assert some file exists there. I wouldn't worry too much about platform-specific extensions and such, the test can be fairly liberal (e.g. check for *any* dll, not just the one platform dll)

**Comment by KalitaAlexey** [comment]
@alexcrichton,
Should `cargo test --example <example>` work if **<example>** is a library?

Am I right about library prefixes and extensions?
* lib, rlib:
  * Any platform: prefix: lib, extension: rlib
* staticlib:
  * Windows: prefix: none, extension: lib
  * Not Windows: prefix: lib, extension: a
* dylib:
  * Windows: prefix: none, extension: dll
  * Not Windows: prefix: lib, extension: so

**Comment by alexcrichton** [comment]
It should do whatever it does today. If it runs unit tests then it should run unit test. If it just builds the artifact then it should just build the artifact.

On OSX the dylib extension is "dylib", but other than that the info you have I believe is correct.

**Comment by KalitaAlexey** [comment]
In the PR `cargo test --example <example>` where `<example>` points to an example of some kind of library doesn't compile.

**Comment by alexcrichton** [comment]
What happens if you point it to an example that's a binary? Libraries should match the same behavior.

**Comment by KalitaAlexey** [comment]
@alexcrichton,
Okay.
It requires changes.
I will update the PR.

**Comment by KalitaAlexey** [comment]
It works.

File: src/cargo/core/manifest.rs [hunk 1]
```diff
@@ -104,18 +104,21 @@ pub enum TargetKind {
...
     CustomBuild,
 }
 
 impl Encodable for TargetKind {
     fn encode<S: Encoder>(&self, s: &mut S) -> Result<(), S::Error> {
         match *self {
-            TargetKind::Lib(ref kinds) => {
-                kinds.iter().map(|k| k.crate_type()).collect()
+            TargetKind::Lib(ref kinds) |
+            // TODO: I am not sure whether it should be encoded like a library or like an example
```
Comment (alexcrichton) [inline src/cargo/core/manifest.rs]: (it's ok to remove this now)

**Comment by alexcrichton** [comment]
Looks like there's a failure on Travis as well?

**Comment by KalitaAlexey** [comment]
@alexcrichton,
Why did it fail?
On my machine all tests had completed without failures.

**Comment by KalitaAlexey** [comment]
I hope it won't fail again.

**Comment by KalitaAlexey** [comment]
@alexcrichton,
Can you include the PR in https://github.com/rust-lang/rust/pull/39199 ?

**Comment by alexcrichton** [comment]
@bors: r+

Looks good, thanks! I think that rollup PR is for rust-lang/rust, not Cargo, so this can't be included.

**Comment by jethrogb** [comment]
Contrary to what's in OP, `cdylib` and `staticlib` are now also supported with this PR. Yay!

**Comment by jethrogb** [comment]
Follow up issue in #3572

**Comment by KalitaAlexey** [comment]
@alexcrichton,
I want to change http://doc.crates.io/guide.html
because it contains "Example executable files go in the examples directory.".
It should be "Example executable or library files go in the examples directory".
How can I change that?

**Comment by alexcrichton** [comment]
@KalitaAlexey the docs are in `src/doc` in this repo, as I think you've found!

**Comment by ehuss** [comment]
I have a minor problem with this change.  `cargo metadata` now outputs `"kind": ["staticlib"]` instead of `"kind": ["example"]` when you have an example library.  I'm working on a tool that uses `cargo metadata` to determine the command-line arguments to use to build a specific target.  Since there is no longer an indication that the target is an example, it is impossible to know that you need to use --example to build.  Should I maybe open a new issue?

**Comment by alexcrichton** [comment]
@ehuss yeah want to open a new issue?

**Comment by KalitaAlexey** [comment]
Hi @ehuss,
I am sorry about that. When I was working on it I didn't know what it was for.

**Comment by alexcrichton** [comment]
@KalitaAlexey since we're seeing some breakage want to send a PR to revert back to the original behavior? We can probably add a new field to the metadata which indicates whether an example is a library or not.

**Comment by KalitaAlexey** [comment]
@alexcrichton,
I will send a PR today (in 3 hours).

**Pull request merged**